gnosis/eth/multicall.py

```diff
@@ -0,0 +1,182 @@
+    success: bool
+    return_data_decoded: Any
+
+
+class MulticallException(Exception):
+    pass
+
+
+class MulticallFunctionFailed(MulticallException):
+    pass
+
+
+class Multicall:
+    ADDRESSES = {
+        EthereumNetwork.MAINNET: '0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696',
+        EthereumNetwork.ARBITRUM: '0x021CeAC7e681dBCE9b5039d2535ED97590eB395c',
+        EthereumNetwork.ARBITRUM_TESTNET: '0xC80E33a6f02CF08557a0CA3D94d1474D73F64Bc1',
+        EthereumNetwork.BINANCE: '0xed386Fe855C1EFf2f843B910923Dd8846E45C5A4',
+        EthereumNetwork.GOERLI: '0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696',
+        EthereumNetwork.KOVAN: '0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696',
+        EthereumNetwork.MATIC: '0xed386Fe855C1EFf2f843B910923Dd8846E45C5A4',
+        EthereumNetwork.MUMBAI: '0xed386Fe855C1EFf2f843B910923Dd8846E45C5A4',
+        EthereumNetwork.RINKEBY: '0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696',
+        EthereumNetwork.ROPSTEN: '0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696',
+        EthereumNetwork.XDAI: '0x08612d3C4A5Dfe2FaaFaFe6a4ff712C2dC675bF7',
+    }
+
+    def __init__(self, ethereum_client: EthereumClient, multicall_contract_address: Optional[ChecksumAddress] = None):
+        self.ethereum_client = ethereum_client
+        self.w3 = ethereum_client.w3
+        ethereum_network = ethereum_client.get_network()
+        address = multicall_contract_address or self.ADDRESSES.get(ethereum_network)
+        if not address:
+            raise EthereumNetworkNotSupported('Multicall contract not available for %s', ethereum_network.name)
+        self.contract = self.get_contract(self.w3, address)
+
+    def get_contract(self, w3: Web3, address: Optional[ChecksumAddress] = None):
+        return w3.eth.contract(address, abi=multicall_v2_abi, bytecode=multicall_v2_bytecode)
+
+    @classmethod
+    def deploy_contract(cls, ethereum_client: EthereumClient, deployer_account: LocalAccount) -> EthereumTxSent:
+        """
+        Deploy contract
+
+        :param ethereum_client:
+        :param deployer_account: Ethereum Account
+        :return: deployed contract address
+        """
+        contract = cls.get_contract(cls, ethereum_client.w3)
+        tx = contract.constructor().buildTransaction({'from': deployer_account.address})
+
+        tx_hash = ethereum_client.send_unsigned_transaction(tx, private_key=deployer_account.key)
+        tx_receipt = ethereum_client.get_transaction_receipt(tx_hash, timeout=120)
+        assert tx_receipt and tx_receipt['status']
+        contract_address = tx_receipt['contractAddress']
+        logger.info("Deployed Multicall V2 Contract %s by %s", contract_address,
+                    deployer_account.address)
+        # Add address to addresses dictionary
+        cls.ADDRESSES[ethereum_client.get_network()] = contract_address
+        return EthereumTxSent(tx_hash, tx, contract_address)
+
+    def _build_payload(self, contract_functions: Sequence[ContractFunction]
```

Comment:
💡  `self` seems to not be used here. maybe we can convert this method to a `@staticmethod`

Reply:
Agree